File: packages/core/src/signing.ts
/**
 * Canonical JSON and operation signing — PROTOCOL.md §8.1 and §12.
 *
 * These live in core rather than crypto because core needs them to validate an
 * op, and crypto already depends on core. Putting them the other way round
 * would make the graph cyclic. Crypto re-exports both, so its public API is
 * unchanged.
 *
 * Everything here is bytes in, bytes out — it knows nothing about operations,
 * which is what lets it be tested without an engine.
 */

import { ed25519 } from "@noble/curves/ed25519.js";
import { sha256 } from "@noble/hashes/sha2.js";

import type { Json } from "./op.js";

/**
 * Prefixed to every signed payload so a signature can never be replayed as
 * valid in some other context.
 */
export const SIGNING_DOMAIN = "tangentfeed/v2/op";

export interface DeviceKey {
  readonly publicKey: Uint8Array;
  readonly privateKey: Uint8Array;
}

/** Canonical JSON per RFC 8785. §8.1. */
export function canonicalJson(value: Json): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return "[" + value.map(canonicalJson).join(",") + "]";
  const keys = Object.keys(value).sort();
  return (
    "{" +
    keys.map((k) => JSON.stringify(k) + ":" + canonicalJson(value[k] as Json)).join(",") +
    "}"
  );
}

export function generateDeviceKey(): DeviceKey {
  const { secretKey, publicKey } = ed25519.keygen();
  return { publicKey, privateKey: secretKey };
}

/**
 * deviceId is the first 16 bytes of SHA-256(publicKey), lowercase hex.
 *
 * 128 bits rather than v0.1's 64: this identifier became a security boundary
 * when it started deciding whose signature counts, and a targeted
 * impersonation at 64 bits is within reach of a determined adversary.
 */
export function deviceIdFromPublicKey(publicKey: Uint8Array): string {
  const digest = sha256(publicKey);
  let out = "";
  for (let i = 0; i < 16; i++) out += digest[i]!.toString(16).padStart(2, "0");
  return out;
}

export function signPayload(payload: Uint8Array, privateKey: Uint8Array): string {
  return base64Encode(ed25519.sign(payload, privateKey));
}

/**
 * Returns false rather than throwing on malformed input. A bad signature from
 * a peer is a routine condition on an open network, not an exceptional one.
 */
export function verifyPayload(
  payload: Uint8Array,
  signature: string,
  publicKey: Uint8Array,
): boolean {
  try {
    const sig = base64Decode(signature);
    if (sig.length !== 64) return false;
    return ed25519.verify(sig, payload, publicKey);
  } catch {
    return false;
  }
}

function base64Encode(bytes: Uint8Array): string {
  if (typeof btoa === "function") {
    let bin = "";
    for (const b of bytes) bin += String.fromCharCode(b);
    return btoa(bin);
  }
  return Buffer.from(bytes).toString("base64");
}

function base64Decode(s: string): Uint8Array {
  if (!/^[A-Za-z0-9+/]*={0,2}$/.test(s)) throw new Error("not base64");
  if (typeof atob === "function") {
    const bin = atob(s);
    return Uint8Array.from(bin, (c) => c.charCodeAt(0));
  }
  return new Uint8Array(Buffer.from(s, "base64"));
}

File: impl/dart/lib/src/signing.dart
/// Operation signing — PROTOCOL.md section 12.
///
/// Ed25519 comes from package:cryptography and is asynchronous; pointycastle,
/// already a dependency for scrypt, has no Ed25519 at all. SHA-256 does come
/// from pointycastle so that deviceId derivation stays synchronous — it is
/// called from constructors and comparisons where a Future would be poison.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:pointycastle/digests/sha256.dart';

/// Prefixed to every signed payload so a signature can never be valid in
/// another context.
const String signingDomain = 'tangentfeed/v2/op';

final Ed25519 _ed25519 = Ed25519();

class DeviceKey {
  const DeviceKey({required this.publicKey, required this.privateKey});

  final Uint8List publicKey;

  /// The 32-byte Ed25519 seed, which is what package:cryptography calls the
  /// private key bytes.
  final Uint8List privateKey;
}

Future<DeviceKey> generateDeviceKey() async {
  final pair = await _ed25519.newKeyPair();
  return DeviceKey(
    publicKey: Uint8List.fromList((await pair.extractPublicKey()).bytes),
    privateKey: Uint8List.fromList(await pair.extractPrivateKeyBytes()),
  );
}

/// deviceId is the first 16 bytes of SHA-256(publicKey), lowercase hex.
///
/// 128 bits rather than v0.1's 64: the identifier became a security boundary
/// when it started deciding whose signature counts.
String deviceIdFromPublicKey(Uint8List publicKey) {
  final digest = SHA256Digest().process(publicKey);
  final buf = StringBuffer();
  for (var i = 0; i < 16; i++) {
    buf.write(digest[i].toRadixString(16).padLeft(2, '0'));
  }
  return buf.toString();
}

Future<String> signPayload(Uint8List payload, Uint8List privateKey) async {
  final pair = await _ed25519.newKeyPairFromSeed(privateKey);
  final signature = await _ed25519.sign(payload, keyPair: pair);
  return base64.encode(signature.bytes);
}

/// Returns false rather than throwing on malformed input: a bad signature from
/// a peer is a routine condition on an open network, not an exceptional one.
Future<bool> verifyPayload(
  Uint8List payload,
  String signature,
  Uint8List publicKey,
) async {
  try {
    final bytes = base64.decode(signature);
    if (bytes.length != 64) return false;
    return await _ed25519.verify(
      payload,
      signature: Signature(
        bytes,
        publicKey: SimplePublicKey(publicKey, type: KeyPairType.ed25519),
      ),
    );
  } catch (_) {
    return false;
  }
}
